src/components/WalletInit/RestoreWallet/WalletCredentialsScreen.js

```diff
@@ -1,57 +1,356 @@
-}: Props) => <WalletForm onSubmit={navigateToWallet} />
+type State = {
+  name: string,
+  password: string,
+  byronAddress: string,
+  shelleyAddress: string,
+  currentDialogStep: restorationDialogSteps,
+  fundedLegacyAddresses: Array<string>,
+  outputAddressInfo: ?{bech32: string, hex: string},
+  balance: ?BigNumber,
+  fee: ?BigNumber,
+  isProcessing: boolean,
+  transferTx: ?TransferTx,
+}
+
+class WalletCredentialsScreen extends React.Component<Props, State> {
+  state = {
+    name: '',
+    password: '',
+    byronAddress: '',
+    shelleyAddress: '',
+    currentDialogStep: RESTORATION_DIALOG_STEPS.CLOSED,
+    fundedLegacyAddresses: [],
+    outputAddressInfo: null,
+    balance: null,
+    fee: null,
+    isProcessing: false,
+    transferTx: null,
+  }
+
+  navigateToWallet = ignoreConcurrentAsync(async (): Promise<void> => {
+    const {name, password} = this.state
+    const {navigation, createWallet} = this.props
+    const phrase = navigation.getParam('phrase')
+    const isShelleyWallet = !!navigation.getParam('isShelleyWallet')
+    await createWallet(name, phrase, password, isShelleyWallet)
+    navigation.navigate(ROOT_ROUTES.WALLET)
+  }, 1000)
+
+  onSubmitWalletCredentials = ({name, password}) => {
+    // TODO: make sure this call is executed only once
+    const self = this
+    this.setState(
+      {
+        name,
+        password,
+        // call on next tick because setState is async
+      },
+      () => self.startWalletRestoration(),
+    )
+  }
+
+  startWalletRestoration = async () => {
+    const {navigation, intl} = this.props
+    const isShelleyWallet = navigation.getParam('isShelleyWallet')
+    if (isShelleyWallet === false) {
+      this.navigateToWallet()
+      return
+    }
+    this.setState({isProcessing: true})
+    const phrase = navigation.getParam('phrase')
+    try {
+      // get first external byron address
+      const byronAddr = await getAddressesFromMnemonics(
+        phrase,
+        displayAddrType,
+        [0],
+      )
+      // get first external group address (bech32)
+      const shelleyAddr = await getGroupAddressesFromMnemonics(
+        phrase,
+        displayAddrType,
+        [0],
+      )
+      this.setState({
+        isProcessing: false,
+        byronAddress: byronAddr[0],
+        shelleyAddress: shelleyAddr[0],
+        currentDialogStep: RESTORATION_DIALOG_STEPS.WALLET_VERIFY,
+      })
+    } catch (e) {
+      this.setState({
+        currentDialogStep: RESTORATION_DIALOG_STEPS.CLOSED,
+        isProcessing: false,
+      })
+      await showErrorDialog(errorMessages.generalError, intl, {
+        message: e.message,
+      })
+    }
+  }
+
+  onConfirmVerify = () =>
+    this.setState({currentDialogStep: RESTORATION_DIALOG_STEPS.CHECK_UPGRADE})
+
+  onBack = () =>
+    this.setState({
+      currentDialogStep: RESTORATION_DIALOG_STEPS.CLOSED,
+    })
+
+  onCheck = async () => {
+    // TODO: there is currently a bug when displaying the UpgradeCheckModal
```

Comment:
TODO?

Reply:
Still todo but not critical IMHO. I haven't figured out why this happens.